diff -urN radiusd-cistron-1.6.4.stock/doc/README.mysql radiusd-cistron-1.6.4/doc/README.mysql
--- radiusd-cistron-1.6.4.stock/doc/README.mysql	Thu Apr 12 17:37:13 2001
+++ radiusd-cistron-1.6.4/doc/README.mysql	Thu Apr 12 17:41:57 2001
@@ -78,6 +78,7 @@
   RadacctID int unsigned DEFAULT '0' NOT NULL auto_increment,
   AcctSessionId varchar(32),
   UserName varchar(32) DEFAULT '' NOT NULL,
+  Realm varchar(32) DEFAULT '' NOT NULL,
   NASIPAddress varchar(15),
   NASPortId varchar(15),
   NASPortType varchar(32),
diff -urN radiusd-cistron-1.6.4.stock/src/mysql.c radiusd-cistron-1.6.4/src/mysql.c
--- radiusd-cistron-1.6.4.stock/src/mysql.c	Thu Apr 12 17:37:13 2001
+++ radiusd-cistron-1.6.4/src/mysql.c	Thu Apr 12 17:40:13 2001
@@ -49,7 +49,7 @@
 	char		*valbuf;
 	VALUE_PAIR *pair;
 	struct stat backup;
-	struct mysqlrec sqlrecord = {"", "", "", "", "", "", 0, "", 0, 0, "", "", "", "", "", 0, 0};
+	struct mysqlrec sqlrecord = {"", "", "", "", "", "", "", 0, "", 0, 0, "", "", "", "", "", 0, 0};
 
 	/* Get timestamp */
 	time((unsigned long*)&sqlrecord.Tstamp);
@@ -68,6 +68,10 @@
 			strncpy(sqlrecord.UserName, pair->strvalue, SQLBIGREC);
 			break;
 
+			case PW_REALM:
+			strncpy(sqlrecord.Realm, pair->strvalue, SQLBIGREC);
+			break;
+
 			case PW_NAS_IP_ADDRESS:
 			ipaddr2str(sqlrecord.NASIPAddress, pair->lvalue);
 			break;
@@ -208,10 +212,11 @@
 
 				while(!feof(mysqlfile)) {
 					if(fread(&sqlrecord, sizeof(struct mysqlrec), 1, mysqlfile) == 1) {;
-						snprintf(querystr, 2048, "insert into %s values (0, '%s', '%s', '%s', '%s', '%s', '%s', %ld, '%s', %ld, %ld, '%s', '%s', '%s', '%s', '%s', %ld, %ld)",
+						snprintf(querystr, 2048, "insert into %s values (0, '%s', '%s', '%s', '%s', '%s', '%s', '%s', %ld, '%s', %ld, %ld, '%s', '%s', '%s', '%s', '%s', %ld, %ld)",
 							sqlTable,
 							sqlrecord.AcctSessionId,
 							sqlrecord.UserName,
+							sqlrecord.Realm,
 							sqlrecord.NASIPAddress,
 							sqlrecord.NASPortId,
 							sqlrecord.NASPortType,
@@ -243,10 +248,11 @@
 		*  current incoming records into the database
 		*/
 
-		snprintf(querystr, 2048, "insert into %s values (0, '%s', '%s', '%s', '%s', '%s', '%s', %ld, '%s', %ld, %ld, '%s', '%s', '%s', '%s', '%s', %ld, %ld)",
+		snprintf(querystr, 2048, "insert into %s values (0, '%s', '%s', '%s', '%s', '%s', '%s', '%s', %ld, '%s', %ld, %ld, '%s', '%s', '%s', '%s', '%s', %ld, %ld)",
 			sqlTable,
 			sqlrecord.AcctSessionId,
 			sqlrecord.UserName,
+			sqlrecord.Realm,
 			sqlrecord.NASIPAddress,
 			sqlrecord.NASPortId,
 			sqlrecord.NASPortType,
diff -urN radiusd-cistron-1.6.4.stock/src/mysql.h radiusd-cistron-1.6.4/src/mysql.h
--- radiusd-cistron-1.6.4.stock/src/mysql.h	Thu Apr 12 17:37:13 2001
+++ radiusd-cistron-1.6.4/src/mysql.h	Thu Apr 12 17:40:13 2001
@@ -22,6 +22,7 @@
 struct mysqlrec {
 	char				AcctSessionId[SQLBIGREC];
 	char				UserName[SQLBIGREC];
+	char				Realm[SQLBIGREC];
 	char				NASIPAddress[SQLLILREC];
 	char				NASPortId[SQLLILREC];
 	char				NASPortType[SQLBIGREC];
