diff -urN radiusd-cistron-1.6.4.stock/src/auth.c radiusd-cistron-1.6.4/src/auth.c
--- radiusd-cistron-1.6.4.stock/src/auth.c	Thu Apr 12 17:26:31 2001
+++ radiusd-cistron-1.6.4/src/auth.c	Thu Apr 12 17:27:11 2001
@@ -395,7 +395,7 @@
 			break;
 		case PW_AUTHTYPE_LDAP:
 #ifdef USELDAP
-			if (ldap_pass(name, string) != 0)
+			if (ldap_pass(name, string, namepair) != 0)
 				result = -1;
 #else
 			log(L_ERR, "%s: LDAP Authentication not available", name);
diff -urN radiusd-cistron-1.6.4.stock/src/files.c radiusd-cistron-1.6.4/src/files.c
--- radiusd-cistron-1.6.4.stock/src/files.c	Mon Aug 21 14:39:12 2000
+++ radiusd-cistron-1.6.4/src/files.c	Thu Apr 12 17:27:11 2001
@@ -1989,7 +1989,7 @@
 	char	buffer[256];
 	char	realm[32];
 	char	hostnm[128];
-	char	opts[5][32];
+	char	opts[6][64];
 	char	*p;
 	int	i;
 	int	lineno = 0;
@@ -2008,10 +2008,10 @@
 			continue;
 		opts[0][0] = opts[1][0] = 0;
 		opts[2][0] = opts[3][0] = 0;
-		opts[4][0] = 0;
-		if (sscanf(buffer, "%31s%127s%31s%31s%31s%31s%31s",
+		opts[4][0] = opts[5][0] = 0;
+		if (sscanf(buffer, "%31s%127s%63s%63s%63s%63s%63s%63s",
 		    realm, hostnm, opts[0], opts[1],
-		    	opts[2], opts[3], opts[4]) < 2) {
+		    	opts[2], opts[3], opts[4], opts[5]) < 2) {
 			log(L_ERR, "%s[%d]: syntax error", file, lineno);
 			return -1;
 		}
@@ -2038,7 +2038,7 @@
 		strNcpy(c->server, hostnm, sizeof(c->server));
 
 		/* Yes we could do this more intelligently */
-		for(i = 0; i < 3; i++) {
+		for(i = 0; i < 5; i++) {
 			if (strcmp(opts[i], "nostrip") == 0)
 				c->striprealm = 0;
 			if (strcmp(opts[i], "hints") == 0)
@@ -2049,6 +2049,8 @@
 				c->auth_port = 0;
 			if (strcmp(opts[i], "trusted") == 0)
 				c->trusted = 0;
+			if (strncmp(opts[i], "basedn", 6) == 0)
+				strNcpy(c->basedn, opts[i] + 7, 63);
 		}
 		c->next = realms;
 		realms = c;
diff -urN radiusd-cistron-1.6.4.stock/src/ldap.c radiusd-cistron-1.6.4/src/ldap.c
--- radiusd-cistron-1.6.4.stock/src/ldap.c	Thu Apr 12 17:26:31 2001
+++ radiusd-cistron-1.6.4/src/ldap.c	Thu Apr 12 17:29:10 2001
@@ -159,11 +159,12 @@
  *************************************************************************/
 
 int
-ldap_pass(char *name, char *passwd)
+ldap_pass(char *name, char *passwd, VALUE_PAIR *namepair)
 {
 	static LDAP *ld;
 	LDAPMessage *result, *msg;
 	char *filter, *dn, *attrs[] = {"uid", NULL};
+	REALM *realm;
 
 	if (use_ldap_auth == 0) {
 		log(L_ERR, "LDAP Auth specified in users file, but not in ldapserver file");
@@ -183,7 +184,10 @@
 
 	filter = make_filter(ldap_filter, name);
 
-	if (ldap_search_s(ld, ldap_basedn, LDAP_SCOPE_SUBTREE, filter, attrs, 1, &result) != LDAP_SUCCESS) {
+	if (namepair->userealm == 'Y')
+		realm = realm_find(namepair->realm != NULL ? namepair->realm : "NULL");
+
+	if (ldap_search_s(ld, namepair->userealm == 'Y' ? realm->basedn : ldap_basedn, LDAP_SCOPE_SUBTREE, filter, attrs, 1, &result) != LDAP_SUCCESS) {
 		ldap_unbind_s(ld);
 		return -1;
 	}
diff -urN radiusd-cistron-1.6.4.stock/src/proxy.c radiusd-cistron-1.6.4/src/proxy.c
--- radiusd-cistron-1.6.4.stock/src/proxy.c	Mon Aug 21 14:39:12 2000
+++ radiusd-cistron-1.6.4/src/proxy.c	Thu Apr 12 17:27:11 2001
@@ -304,6 +304,13 @@
 	if (strcmp(realm->server, "LOCAL") == 0) {
 		/* Same size */
 		strcpy(namepair->strvalue, saved_username);
+
+		realmname = strrchr(namepair->strvalue, '@');
+		if (realmname != NULL) {
+			strncpy(namepair->realm, realmname + 1, 63);
+			namepair->userealm = 'Y';
+		}
+
 		if (realm->striprealm &&
 		    ((realmname = strrchr(namepair->strvalue, '@')) != NULL)) {
 			*realmname = 0;
diff -urN radiusd-cistron-1.6.4.stock/src/radiusd.h radiusd-cistron-1.6.4/src/radiusd.h
--- radiusd-cistron-1.6.4.stock/src/radiusd.h	Thu Apr 12 17:26:31 2001
+++ radiusd-cistron-1.6.4/src/radiusd.h	Thu Apr 12 17:27:38 2001
@@ -70,6 +70,8 @@
 	UINT4			lvalue;
 	int			operator;
 	char			strvalue[AUTH_STRING_LEN];
+	char		userealm;
+	char		realm[64];
 	struct value_pair	*next;
 } VALUE_PAIR;
 
@@ -116,6 +118,7 @@
 typedef struct realm {
 	char			realm[64];
 	char			server[64];
+	char			basedn[128];
 	UINT4			ipaddr;
 	int			auth_port;
 	int			acct_port;
@@ -291,5 +294,5 @@
 #define MAX_AUTH_QUERY_LEN 256
 
 int radldap_init(void);
-int ldap_pass(char *, char *);
+int ldap_pass(char *, char *, VALUE_PAIR *);
 #endif
